### PR TITLE
Wait for healthy extension server before registering APIService, handle ServiceUnavailable errors

### LINKED ISSUE #58642: an error on the server ("missing route (service \"sample-api\" not found)") has prevented the request from succeeding (get flunders.wardle.k8s.io)
`Jan 22 22:15:31.659: Couldn't delete ns: "e2e-tests-statefulset-fg2hp": an error on the server ("missing route (service \"sample-api\" not found)") has prevented the request from succeeding (get flunders.wardle.k8s.io) (&errors.StatusError{ErrStatus:v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:""}, Status:"Failure", Message:"an error on the server (\"missing route (service \\\"sample-api\\\" not found)\") has prevented the request from succeeding (get flunders.wardle.k8s.io)", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc4216bc960), Code:500}})`

* https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/20776
* https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/20777
* https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/20775
* https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/20774
* https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/20772
* ... A lot of them

@kubernetes/sig-api-machinery-bugs 
/cc @cheftako

### PR DESCRIPTION
**Pull request by liggitt**
fixes #58642 
followup to #58070

* Because a registered APIService appears in discovery immediately, we should wait until the backing deployment is healthy before exposing it
* In e2e hasRemainingContent(), add ServiceUnavailable to the types of errors we tolerate when looking for remaining content.
* In proxy handler, return a ServiceUnavailable error if the referenced service cannot be resolved
```release-note
NONE
```

### CODE REVIEW

File: test/e2e/apimachinery/aggregator.go [hunk 1]
```diff
@@ -291,6 +291,12 @@ func TestSampleAPIServer(f *framework.Framework, image string) {
 	})
 	framework.ExpectNoError(err, "creating role binding %s:sample-apiserver to access configMap", namespace)
 
+	// Wait for the extension apiserver to be up and healthy
+	// kubectl get deployments -n <aggregated-api-namespace> && status == Running
+	// NOTE: aggregated apis should generally be set up in there own namespace (<aggregated-api-namespace>). As the test framework
+	// is setting up a new namespace, we are just using that.
+	err = framework.WaitForDeploymentComplete(client, deployment)
```
Comment (sttts) [inline test/e2e/apimachinery/aggregator.go]: lgtm. Not sure it's the source for the flakes, but it could be.

Comment (weekface) [reply test/e2e/apimachinery/aggregator.go]: Do we need add a `framework.ExpectNoError` call to ensure `err` is nil?

Comment (liggitt) [reply test/e2e/apimachinery/aggregator.go]: yes, added

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -1200,6 +1200,10 @@ func hasRemainingContent(c clientset.Interface, clientPool dynamic.ClientPool, n
 			if apierrs.IsMethodNotSupported(err) || apierrs.IsNotFound(err) || apierrs.IsForbidden(err) {
 				continue
 			}
+			// skip unavailable servers
```
Comment (sttts) [inline test/e2e/framework/util.go]: maybe add a comment why we add this

**Comment by sttts** [comment]
Only a nit. Otherwise lgtm.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 3]
```diff
@@ -120,7 +120,8 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	location.Scheme = "https"
 	rloc, err := r.serviceResolver.ResolveEndpoint(handlingInfo.serviceNamespace, handlingInfo.serviceName)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("missing route (%s)", err.Error()), http.StatusInternalServerError)
+		glog.Errorf("error resolving %s/%s: %v", handlingInfo.serviceName, handlingInfo.serviceNamespace, err)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go]: ns/name

**Comment by Random-Liu** [comment]
@liggitt Thanks for the fix! :)

**Pull request merged**